virttest/qemu_vm.py

```diff
@@ -4252,10 +4252,21 @@ def _shell_reboot(session, timeout):
         if self.params.get("mac_changeable") == "yes":
             utils_net.update_mac_ip_address(self)
 
-        if serial:
-            return self.wait_for_serial_login(timeout=(timeout - shutdown_dur))
-        return self.wait_for_login(nic_index=nic_index,
-                                   timeout=(timeout - shutdown_dur))
+        end = time.time() + timeout - shutdown_dur
+        while True:
+            try:
+                _timeout = max(10, end - time.time())
+                if serial:
+                    return self.wait_for_serial_login(timeout=_timeout)
+                return self.wait_for_login(nic_index=nic_index,
+                                           timeout=_timeout)
+            except virt_vm.VMError as details:
```

Comment:
I've checked that `virt_vm.VMError` could be raised at the following three places.
https://github.com/avocado-framework/avocado-vt/blob/261283289cbc4b1d77a9992f86dac112214aa815/virttest/virt_vm.py#L1364
https://github.com/avocado-framework/avocado-vt/blob/261283289cbc4b1d77a9992f86dac112214aa815/virttest/virt_vm.py#L1177
https://github.com/avocado-framework/avocado-vt/blob/261283289cbc4b1d77a9992f86dac112214aa815/virttest/virt_vm.py#L1180

About the first two I think we can remove them directly and that would be harmless I guess. And for the last one, the error should be raised because we've waited long enough at that time. Thus, we don't have to have this commit to run the `wait_for_x` things in a loop, and it can handle the login case as well.

Reply:
Well I'd like to avoid unnecessary exceptions and I'm not even sure how to realistically ignore some of the exceptions of the same type. Anyway not raising it shouldn't be a problem since we do log the details and, provided we waited long enough, the loop finishes and results in "Unable to login ..." exception.